src/upstream_service.cpp

```diff
@@ -0,0 +1,294 @@
+          << bytes_parsed << " bytes out of " << sampling_rate_string.size()
+          << "): " << sampling_rate_string;
+    throw std::invalid_argument(error.str());
+  }
+
+  return sampling_rate;
+} catch (const std::logic_error& error) {
+  throw std::invalid_argument(std::string("Unable to parse sampling rate (error: ") +
+                              error.what() + ") from the following text: " + std::string(text));
+}
+
+// Return an `UpstreamService` parsed from the specified `text`, or throw a
+// `std::invalid_argument` if an error occurs.
+UpstreamService deserialize(ot::string_view text) {
+  UpstreamService result;
+  const auto end = text.end();
+
+  // .service_name
+  auto iter = text.begin();
+  auto next = std::find(iter, end, '|');
+  deserializeServiceName(result.service_name, range(iter, next));
+
+  // `UpstreamService` fields are "|" separated.  Every time we go to the next
+  // field, we have to make sure that there was a "|" before it (otherwise
+  // there aren't enough fields).
+  const auto advanceTo = [&](const char* field_name_for_diagnostic) {
+    if (next == end) {
+      std::string error;
+      error += "Missing fields in UpstreamService; no ";
+      error += field_name_for_diagnostic;
+      error += " to decode. Offending text: ";
+      error += text;
+      throw std::invalid_argument(error);
+    }
+    iter = next + 1;
+    next = std::find(iter, end, '|');
+  };
+
+  advanceTo("sampling_priority");
+  result.sampling_priority = deserializeSamplingPriority(range(iter, next));
+
+  advanceTo("sampling_mechanism");
+  result.sampling_mechanism = deserializeSamplingMechanism(range(iter, next));
+
+  advanceTo("sampling_rate");
+  result.sampling_rate = deserializeSamplingRate(range(iter, next));
+
+  // Put any remaining fields into `result.unknown_fields`.
+  while (next != end) {
+    iter = next + 1;
+    next = std::find(iter, end, '|');
+    result.unknown_fields.push_back(range(iter, next));
+  }
+
+  return result;
+}
+
+// Append to the specified `output` the serialization of the specified
+// `upstream_service`.
+void serialize(std::string& output, const UpstreamService& upstream_service) {
+  output += base64_rfc4648_unpadded::encode(upstream_service.service_name);
+
+  output += '|';
+  output += std::to_string(int(upstream_service.sampling_priority));
+
+  output += '|';
+  output += std::to_string(upstream_service.sampling_mechanism);
+
+  output += '|';
+  appendSamplingRate(output, upstream_service.sampling_rate);
+
+  for (const std::string& value : upstream_service.unknown_fields) {
+    output += '|';
+    output += value;
+  }
+}
+
+// This function is used in `operator==`, below.
+std::string serializeSamplingRate(double value) {
+  std::string result;
+  appendSamplingRate(result, value);
+  return result;
+}
+
+}  // namespace
+
+void appendSamplingRate(std::string& destination, double value) {
+  // NaNs serialize as an empty string.
+  if (std::isnan(value)) {
+    return;
+  }
```

Comment:
Not quite related to this code, but: do create a fuzzing prototype. Maybe the code will break and scare you into a more conservative solution.

Reply:
Fuzz the encoder, too, because the service name is untrusted (even though we build the structure).